pkg/collector/corechecks/gpu/gpu.go

```diff
@@ -0,0 +1,206 @@
+	statProcessors map[uint32]*StatsProcessor
+}
+
+// Factory creates a new check factory
+func Factory() optional.Option[func() check.Check] {
+	return optional.NewOption(newCheck)
+}
+
+func newCheck() check.Check {
+	return &Check{
+		CheckBase:      core.NewCheckBase(CheckName),
+		config:         &CheckConfig{},
+		statProcessors: make(map[uint32]*StatsProcessor),
+	}
+}
+
+// Parse parses the check configuration
+func (m *CheckConfig) Parse(data []byte) error {
+	return yaml.Unmarshal(data, m)
+}
+
+// Cancel cancels the check
+func (m *Check) Cancel() {
+	ret := nvml.Shutdown()
+	if ret != nvml.SUCCESS && ret != nvml.ERROR_UNINITIALIZED {
+		log.Warnf("Failed to shutdown NVML: %v", nvml.ErrorString(ret))
+	}
+}
+
+// Configure parses the check configuration and init the check
+func (m *Check) Configure(senderManager sender.SenderManager, _ uint64, config, initConfig integration.Data, source string) error {
+	if err := m.CommonConfigure(senderManager, initConfig, config, source); err != nil {
+		return err
+	}
+	if err := m.config.Parse(config); err != nil {
+		return fmt.Errorf("gpu check config: %s", err)
+	}
+
+	return nil
+}
+
+func (m *Check) ensureInitialized() error {
+	var err error
+
+	if m.sysProbeUtil == nil {
+		m.sysProbeUtil, err = processnet.GetRemoteSystemProbeUtil(
+			pkgconfigsetup.SystemProbe().GetString("system_probe_config.sysprobe_socket"),
+		)
+		if err != nil {
+			return fmt.Errorf("sysprobe connection: %s", err)
+		}
+	}
+
+	if m.timeResolver == nil {
+		m.timeResolver, err = sectime.NewResolver()
+		if err != nil {
+			return fmt.Errorf("cannot create time resolver: %s", err)
+		}
+	}
+	return nil
+}
+
+// ensureProcessor ensures that there is a stats processor for the given key
+func (m *Check) ensureProcessor(key *model.StreamKey, snd sender.Sender, gpuThreads int, checkDuration time.Duration) {
+	if _, ok := m.statProcessors[key.Pid]; !ok {
+		m.statProcessors[key.Pid] = &StatsProcessor{
+			key: key,
+		}
+	}
+
+	m.statProcessors[key.Pid].totalThreadSecondsUsed = 0
+	m.statProcessors[key.Pid].sender = snd
+	m.statProcessors[key.Pid].gpuMaxThreads = gpuThreads
+	m.statProcessors[key.Pid].measuredInterval = checkDuration
+	m.statProcessors[key.Pid].timeResolver = m.timeResolver
+	m.statProcessors[key.Pid].lastCheck = m.lastCheckTime
+}
+
+// Run executes the check
+func (m *Check) Run() error {
+	if err := m.ensureInitialized(); err != nil {
+		return err
+	}
+
+	gpuDevices, err := getGPUDevices()
+	if err != nil {
+		return err
+	}
+
+	if len(gpuDevices) == 0 {
+		return fmt.Errorf("no GPU devices found")
+	}
+
+	data, err := m.sysProbeUtil.GetCheck(sysconfig.GPUMonitoringModule)
+	if err != nil {
+		return fmt.Errorf("cannot get data from system-probe: %s", err)
+	}
+	now := time.Now()
+
+	var checkDuration time.Duration
+	// mark the check duration as close to the actual check as possible
+	if !m.lastCheckTime.IsZero() {
+		checkDuration = now.Sub(m.lastCheckTime)
+	}
```

Comment:
It's needed to calculate the interval for which we have data for the kernels, so that we can compute utilization in terms of thread-seconds used divided by thread-seconds available. System-probe will flush the data after it's sent to the agent, so we mark our times as close to that event as possible.